Lib/test/test_gc.py

```diff
@@ -1082,6 +1082,37 @@ def __del__(self):
         gc.collect()
         self.assertTrue(collected)
 
+    def test_traverse_frozen_objects(self):
+        # See GH-126312: Objects that were not frozen could traverse over
+        # a frozen object on the free-threaded build, which would cause
+        # a negative reference count.
+        import textwrap
+
+        x = [1, 2, 3]
+        gc.freeze()
+        y = [x]
+        y.append(y)
+        del y
+        gc.collect()
+        gc.unfreeze()
+
+        source = textwrap.dedent("""
+        import gc
```

Comment:
It's not clear to me if this needs to be executed in a subprocess (via `assert_python_ok`). If it does, then keep the code simple: don't wrap it in a `unittest.TestCase`:

```python
source = textwrap.dedent("""
    import gc
    gc.freeze()
    gc.collect()
    gc.unfreeze()
""")
```

Reply:
See my comment in the original issue: https://github.com/python/cpython/issues/126312#issuecomment-2453166963

I'm triggering the `visit_decref` problem with the code above, and then I'm triggering the deferred reference counting problem (via `visit_stackref`) in this snippet. I'm not totally sure how to trigger it in the wild, but using it in `unittest` reliably crashes as a result of it for me. I'm OK with removing this though, since the part above will implicitly test for this anyway.

Reply:
Ok, that's fine. Please leave a code comment that the `unittest.TestCase` is an important part of the reproducer.